File: Python/ByTier/E/1920_E_Counting_Binary_Strings.py
# Problem: CF 1920 E - Counting Binary Strings
# https://codeforces.com/contest/1920/problem/E

"""
Algorithm: Dynamic Programming with State Transition
Purpose: Count binary strings with exactly n good substrings and max good substring length k.
Time Complexity: O(n^2 * k) per test case
Space Complexity: O(n^2) for the DP array

The problem is solved using dynamic programming where dp[i][j] represents the number of ways to form 
i good substrings ending with a good substring of length j. A good substring is defined as one that 
contains exactly one '1'.

The recurrence builds valid strings by extending existing ones with substrings of length 1 to k,
ensuring that no good substring exceeds length k.

Notes:
- The state dp[i][j] stores the number of valid binary strings with exactly i good substrings,
  ending with a good substring of length j.
- A good substring of length l contributes l positions to the count of good substrings.
"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def f(u, v):
    # Helper function to map 2D state (u, v) to 1D array index
    return u * (n + 1) + v


t = int(input())
ans = []
mod = 998244353
for _ in range(t):
    n, k = map(int, input().split())
    # dp array for dynamic programming
    dp = [0] * ((n + 1) * (n + 1))
    
    # Initialize base case: 1 good substring ending with substring of length 1
    for i in range(1, n + 1):
        dp[i] = 1
    
    # Iterate through all possible counts of good substrings
    for i in range(n):
        # For each possible ending length of good substring
        for j in range(1, n + 1):
            # Skip if this state is unreachable
            if not dp[f(i, j)]:
                continue
            u = dp[f(i, j)] % mod  # Current number of ways to reach this state
            # Try extending with good substrings of length 1 to k-j+1
            for l in range(1, k - j + 2):
                v = i + j * l  # New number of good substrings after extension
                if v > n:
                    break
                # Accumulate the number of ways to reach (v, l) state
                dp[f(v, l)] += u
    
    # Sum all valid final states (ending with good substring of length 1)
    ans0 = 0
    for i in range(f(n, 1), f(n + 1, 0)):
        ans0 += dp[i]
    ans0 %= mod
    ans.append(ans0)

sys.stdout.write("\n".join(map(str, ans)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1920/1920_E_Counting_Binary_Strings.cpp
// Problem: CF 1920 E - Counting Binary Strings
// https://codeforces.com/contest/1920/problem/E

/*
 * Problem: Counting Binary Strings
 * 
 * Algorithm:
 * - Dynamic Programming approach to count valid binary strings.
 * - State definition: dp[i][j] represents the number of ways to form exactly i good substrings 
 *   with the last good substring having length j.
 * - The transition is based on how we can extend the string by appending 0s or 1s such that we 
 *   maintain the constraint that no good substring exceeds length k.
 * 
 * Time Complexity: O(n * k * n) where n is the number of good substrings and k is the max length,
 *                  which simplifies to O(n^2 * k) in worst case.
 * Space Complexity: O(n^2) due to the 2D DP table.
 */

#include <stddef.h>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;
typedef long long ll;
typedef pair<int, int> ii;
typedef pair<ll, ll> pll;
typedef vector<int> vi;
typedef vector<ii> vii;
typedef vector<ll> vll;
typedef vector<pll> vpll;
#define PB push_back
#define PF push_front
#define PPB pop_back
#define PPF pop_front
#define X first
#define Y second
#define MP make_pair
#define all(x) (x).begin(), (x).end()
const int mod = 998244353;
const int inf = 1e9 + 7;
const ll INF = 1e18 + 7;
const int logo = 20;
const int MAXN = 2507;
const int off = 1 << logo;
const int trsz = off << 1;
const int dx[] = {1, -1, 0, 0};
const int dy[] = {0, 0, -1, 1};

void ad(int &a, int b) {
  a += b;
  if (a >= mod)
    a -= mod;
}

int dp[MAXN][MAXN];

void solve() {
  int n, k;
  cin >> n >> k;
  
  // Initialize base case: dp[0][j] = 1 for all j, meaning zero good substrings
  for (int i = 0; i <= n; i++) {
    for (int j = 0; j <= n; j++)
      dp[i][j] = (i == 0);
  }

  int ans = 0;
  
  // Iterate over all possible lengths of the string s (from 1 to n)
  for (int s = 1; s <= n; s++) {
    // Iterate over all possible lengths of the last good substring (from 1 to k)
    for (int cr = 1; cr <= k; cr++) {
      // Iterate over all possible lengths of previous segments to form current segment
      for (int ls = 1; ls + cr - 1 <= k and ls * cr <= s; ls++) {
        // Add transitions: dp[s][cr] += dp[s - ls * cr][ls]
        ad(dp[s][cr], dp[s - ls * cr][ls]);
      }
      // If we've reached the total number of required goods substrings, add to answer
      if (s == n)
        ad(ans, dp[s][cr]);
    }
  }
  
  cout << ans << "\n";
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  int tt = 1;
  cin >> tt;
  while (tt--)
    solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/